File: convert-sorted-list-to-binary-search-tree/convert-sorted-list-to-binary-search-tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        def rec(lo, hi):
            if lo > hi: return
            nonlocal sorted_list
            mid = int((lo+hi)/2)
            node = TreeNode(sorted_list[mid])
            node.left = rec(lo, mid-1)
            node.right = rec(mid+1, hi)
            return node
        
        sorted_list = []
        while head:
            sorted_list.append(head.val)
            head = head.next

        return rec(0, len(sorted_list)-1)

File: convert-sorted-list-to-binary-search-tree/convert-sorted-list-to-binary-search-tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        if(head == null) return null;
        ListNode mid = head, midPrev = null, fast = head;
        for(fast = head; fast != null && fast.next != null; fast = fast.next.next) {
            midPrev = mid;
            mid = mid.next;
        }
        if(mid == head) return new TreeNode(head.val);
        midPrev.next = null;
        if(mid.next != null)
            return new TreeNode(mid.val, sortedListToBST(head), sortedListToBST(mid.next));
        else
            return new TreeNode(mid.val, sortedListToBST(head), null);
    }
    
    void printList(ListNode a) {
        for(; a != null; a = a.next) p.accept(a.val + " ");
        p.accept("\n");
    }
    
    Consumer<Object> p;
    public Solution() {
        p = o -> System.out.print(o);
    }
}